share/system/cgroup_linux.go

```diff
@@ -332,76 +332,147 @@ func getCgroupPath_cgroup_v2(pid int) (string, error) {
+func CgroupRelativePathFix(path string) (string) {
+	if strings.Contains(path, "/..") {
+		// I can't find the documentation why the cgroup file will hold relative paths. For now, I'm applying these
+		// fixes as we encounter them. Documentation on this behaviour would make this code more robust.
+		// Example: 0::/../../kubepods-besteffort-poddee9029c_408f_4466_811d_43eea3042395.slice/docker-a737350ff4843bb79debc4e2dc98f0b1b11d40f814ea4303d9167dd70c314b95.scope
+
+		path  = filepath.Clean(path)
+		if strings.Contains(path, "kubepods-pod") {
+			path = filepath.Join("/kubepods.slice", path)
+		} else if strings.Contains(path, "kubepods-besteffort") {
+			path = filepath.Join("/kubepods.slice/kubepods-besteffort.slice", path)
+		} else if strings.Contains(path, "kubepods-burstable") {
+			path = filepath.Join("/kubepods.slice", path)
+		}
+	}
+	return path
+}
+
+// cgroup v2 is collected inside an unified file folder
+func (s *SystemTools) getCgroupMetricsPath(pid int, subsystem string) (string, error) {
+	var path string
+
+	if pid == 0 { // self
+		path = "/proc/self/cgroup"
+	} else {
+		path = filepath.Join(s.procDir, strconv.Itoa(pid), "cgroup")
 	}
 
-	// However, the k8s POD does not have those subsystem folders
-	path = filepath.Join(s.procDir, strconv.Itoa(pid), "cgroup")
 	f, err := os.Open(path)
 	if err != nil {
+		log.WithFields(log.Fields{"path": path, "err": err}).Warning("cgroup cannot be read, stats cannot be found")
 		return "", err
 	}
-	defer f.Close()
+	defer 	f.Close()
 
-	scanner := bufio.NewScanner(f)
-	for scanner.Scan() {
-		if strings.Contains(scanner.Text(), subsystem) {
-			tokens := strings.Split(scanner.Text(), ":")
-			length := len(tokens)
-			if length == 3 {
-				path := filepath.Join(s.cgroupDir, subsystem, tokens[2])
-				if _, err := os.Stat(path); os.IsNotExist(err) {
-					// Rancher on RancherOS, tokens[2] is /docker/xxxx/docker/yyyy
-					dirs := strings.Split(tokens[2], "/")
-					if n := len(dirs); n > 2 {
-						return filepath.Join(s.cgroupDir, subsystem, dirs[n-2], dirs[n-1]), nil
-					}
-				} else {
-					// Most other systems
-					return path, nil
-				}
-			} else if length > 3 {
-				// containerd 1.4.4: it uses ":" as separators.
-				// 4:memory:/system.slice/containerd.service/kubepods-besteffort-pod2105e389_7471_476e_a50f_6074cef29bbd.slice:cri-containerd:76b9bfe6d5506......
-				path := filepath.Join(s.cgroupDir, subsystem)
-				for i := 2; i < len(tokens); i++ {
-					if i == 2 {
-						path += tokens[i]
-					} else {
-						path += ":" + tokens[i]
-					}
-				}
-				return path, nil
-			}
-			break
-		}
-	}
+	// shouldn't reuse path, doesn't make sense
+	cpath, err := getStatsPathFromCgroupFile(f, subsystem)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"pid": pid,
+			"path": cpath,
+			"s.cgroupDir": s.cgroupDir,
+			"s.procDir": s.procDir,
+			"s.cgroupVersion": s.cgroupVersion,}).
+			Error("Cgroup file could not be parsed, fallback used")
```

Comment:
For debugging the field, the error will dump enough information to help someone to jump on the node and then inspect the container's mounts and determine what went wrong.